File: 한성준/week8/graph_traversal/13023.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, m;
vector<vector<int>> adj;
vector<bool> visited;

bool dfs(int u, int depth) {
    visited[u] = true;

    if (depth == 5) {
        return true;
    }

    for (int v : adj[u]) {
        if (!visited[v]) {
            if (dfs(v, depth + 1)) {
                return true;
            }
        }
    }

    visited[u] = false; // 백트래킹
    return false;
}

int main() {
    cin >> n >> m;

    adj.resize(n);
    for (int i = 0; i < m; ++i) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    for (int i = 0; i < n; ++i) {
        visited.assign(n, false);
        if (dfs(i, 1)) {
            cout << 1 << endl;
            return 0;
        }
    }

    cout << 0 << endl;

    return 0;
}

File: 홍연일/8주차/13023.py
\
import sys
sys.setrecursionlimit(1000000)

# Baekjoon 13023 - ABCDE
N, M = map(int, sys.stdin.readline().split())
adj = [[] for _ in range(N)]
for _ in range(M):
    a, b = map(int, sys.stdin.readline().split())
    adj[a].append(b)
    adj[b].append(a)

visited = [False] * N
found = False

def dfs(v: int, depth: int):
    global found
    if found:
        return
    if depth == 5:
        found = True
        return
    visited[v] = True
    for nxt in adj[v]:
        if not visited[nxt]:
            dfs(nxt, depth + 1)
    visited[v] = False

for i in range(N):
    dfs(i, 1)
    if found:
        break

print(1 if found else 0)
